luarules/gadgets/game_team_power_watcher.lua

```diff
@@ -0,0 +1,294 @@
+
+
+function gadget:GetInfo()
+	return {
+		name = "Team Power Watcher",
+		desc = "Tracks power of individual and total units per team. To be used for PvE dynamic difficulty and library functions to assertain aspects of game progression",
+		author = "SethDGamre",
+		date = "2024-07-12",
+		license = "None",
+		layer = 1,
+		enabled = true
+	}
+end
+
+if gadgetHandler:IsSyncedCode() then
+
+	local spGetGameFrame          = Spring.GetGameFrame
+	local spSetUnitRulesParam     = Spring.SetUnitRulesParam
+	local spGetUnitRulesParam = Spring.GetUnitRulesParam
+	local spGetTeamResources = Spring.GetTeamResources --(teamID, "metal"|"energy") return nil | currentLevel
+	local spGetUnitDefID        = Spring.GetUnitDefID
+
+	local spGetTeamList			= Spring.GetTeamList
+	local spGetUnitTeam 		= Spring.GetUnitTeam
+
+	local spGetGameSeconds = Spring.GetGameSeconds
+
+	local random = math.random
+	local math_min = math.min
+	local sin    = math.sin
+	local cos    = math.cos
+
+	local GAME_SPEED = Game.gameSpeed
+
+	local unitsWithPower = {}
+	local teamList = spGetTeamList()
+	local neutralTeamNumber
+	local teamPowers = {}
+	local highestTeamPower = {}
+    local averageTeamPower = 0
+    local averageAlliedTeamPower = 0
+    local averageHumanTeamPower = 0
+    local averageTechGuesstimate = 0
+    local averageAlliedTechGuesstimate = 0
+    local averageHumanTechGuesstimate = 0
+    local peakTeamPowers = {}
+
+    --AI team lists
+    local scavTeam
+    local raptorTeam
+    local AIteams = {}
+    local humanTeams = {}
+
+
+    --used for tech level guesstimate functions
+    local powerThresholds = {
+        {threshold = 9000, techLevel = 1},
+        {threshold = 45000, techLevel = 1.5},
+        {threshold = 90000, techLevel = 2},
+        {threshold = 230000, techLevel = 2.5},
+        {threshold = 350000, techLevel = 3},
+        {threshold = 475000, techLevel = 3.5},
+        {threshold = 600000, techLevel = 4},
+        {threshold = 725000, techLevel = 4.5}
+    }
+
+
+--decipher human vs ai vs neutral vs defense mode ai's 
+    for _, teamID in ipairs(teamList) do
+		local teamLuaAI = Spring.GetTeamLuaAI(teamID)
+		if (teamLuaAI and string.find(teamLuaAI, "ScavengersAI")) then
+			scavTeam = teamID
+        elseif (teamLuaAI and string.find(teamLuaAI, "RaptorsAI")) then
+            raptorTeam = teamID
+        elseif select (4, Spring.GetTeamInfo(teamID, false)) then
+            AIteams[teamID] = true
+        elseif teamID == tonumber(teamList[#teamList]) then
+            neutralTeamNumber = teamID
+        else
+			humanTeams[teamID] = true
+		end
+	end
+
+--assign all teams power and peak power of 0 to prevent nil errors
+    for _, teamNumber in ipairs(teamList) do
+        teamPowers[teamNumber] = 0
+        peakTeamPowers[teamNumber] = 0
+    end
+
+
+
+	function gadget:UnitFinished(unitID, unitDefID, unitTeam)
+		if UnitDefs[unitDefID].power then
```

Comment:
redundant check, every unit has power

Reply:
Including seagulls and crabs?

Reply:
yes, even commander hats.

Reply:
commander hats... most powerful of all

I'll do as you ask